libs/platform/ebpf_bitmap.c

```diff
@@ -0,0 +1,134 @@
+    size_t bit_count;
+    size_t data[1];
+} ebpf_bitmap_t;
+
+typedef struct _ebpf_bitmap_cursor_internal
+{
+    ebpf_bitmap_t* bitmap;
+    size_t index;         // Current block being searched.
+    size_t current_block; // Copy of the current block.
+} ebpf_bitmap_cursor_internal_t;
+
+C_ASSERT(sizeof(ebpf_bitmap_cursor_internal_t) == sizeof(ebpf_bitmap_cursor_t));
+
+#define BITS_IN_SIZE_T (sizeof(size_t) * 8)
+#define BIT_COUNT_TO_BLOCK_COUNT(X) ((X + BITS_IN_SIZE_T - 1) / BITS_IN_SIZE_T)
+
+size_t
+ebpf_bitmap_size(size_t bit_count)
+{
+    return EBPF_OFFSET_OF(ebpf_bitmap_t, data) + BIT_COUNT_TO_BLOCK_COUNT(bit_count) * sizeof(size_t);
+}
+
+void
+ebpf_bitmap_initialize(_Out_ ebpf_bitmap_t* bitmap, size_t bit_count)
+{
+    bitmap->bit_count = bit_count;
+    memset(bitmap->data, 0, BIT_COUNT_TO_BLOCK_COUNT(bit_count) * sizeof(size_t));
+}
+
+bool
+ebpf_bitmap_set_bit(_Inout_ ebpf_bitmap_t* bitmap, size_t index, bool interlocked)
+{
+    volatile int64_t* block = (volatile int64_t*)(bitmap->data + index / BITS_IN_SIZE_T);
+    uint8_t position_within_block = index % BITS_IN_SIZE_T;
+
+    if (interlocked) {
+        return _interlockedbittestandset64(block, position_within_block);
+    } else {
+        return _bittestandset64((int64_t*)block, position_within_block);
+    }
+}
+
+bool
+ebpf_bitmap_reset_bit(_Inout_ ebpf_bitmap_t* bitmap, size_t index, bool interlocked)
+{
+    volatile int64_t* block = (volatile int64_t*)(bitmap->data + index / BITS_IN_SIZE_T);
+    uint8_t position_within_block = index % BITS_IN_SIZE_T;
+
+    if (interlocked) {
+        return _interlockedbittestandreset64(block, position_within_block);
+    } else {
+        return _bittestandreset64((int64_t*)block, position_within_block);
+    }
+}
+
+bool
+ebpf_bitmap_test_bit(_In_ const ebpf_bitmap_t* bitmap, size_t index)
+{
+    return _bittest64((const int64_t*)&(bitmap->data[index / BITS_IN_SIZE_T]), index % BITS_IN_SIZE_T);
+}
+
+void
+ebpf_bitmap_start_forward_search(_In_ ebpf_bitmap_t* bitmap, _Out_ ebpf_bitmap_cursor_t* cursor)
+{
+    ebpf_bitmap_cursor_internal_t* internal_cursor = (ebpf_bitmap_cursor_internal_t*)cursor;
+    internal_cursor->bitmap = bitmap;
+    internal_cursor->index = 0;
+    internal_cursor->current_block = bitmap->data[internal_cursor->index / BITS_IN_SIZE_T];
+}
+
+void
+ebpf_bitmap_start_reverse_search(_In_ ebpf_bitmap_t* bitmap, _Out_ ebpf_bitmap_cursor_t* cursor)
+{
+    ebpf_bitmap_cursor_internal_t* internal_cursor = (ebpf_bitmap_cursor_internal_t*)cursor;
+    internal_cursor->bitmap = bitmap;
+    internal_cursor->index = bitmap->bit_count - 1;
+    internal_cursor->current_block = bitmap->data[internal_cursor->index / BITS_IN_SIZE_T];
+}
+
+size_t
+ebpf_bitmap_forward_search_next_bit(_Inout_ ebpf_bitmap_cursor_t* cursor)
+{
+    ebpf_bitmap_cursor_internal_t* internal_cursor = (ebpf_bitmap_cursor_internal_t*)cursor;
+    while (internal_cursor->index < internal_cursor->bitmap->bit_count) {
+        DWORD next_bit = 0;
```

Comment:
Hmm, that's odd because the header says:
```
_Success_(return!=0)
BOOLEAN
_BitScanForward64 (
    _Out_ DWORD *Index,
    _In_ DWORD64 Mask
    );

_Success_(return!=0)
BOOLEAN
_BitScanReverse64 (
    _Out_ DWORD *Index,
    _In_ DWORD64 Mask
    );
```